src/tools/kotlin-refinement-generator.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * @license
+ * Copyright (c) 2020 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Op} from '../runtime/manifest-ast-nodes.js';
+import {Dictionary} from '../runtime/hot.js';
+import {Schema} from '../runtime/schema.js';
+import {escapeIdentifier, typeFor, defaultValFor} from './kotlin-codegen-shared.js';
+import {RefinementExpressionVisitor, BinaryExpression, UnaryExpression, FieldNamePrimitive, QueryArgumentPrimitive, BuiltIn, NumberPrimitive, BooleanPrimitive, TextPrimitive} from '../runtime/refiner.js';
+
+// The variable name used for the query argument in generated Kotlin code.
+const KOTLIN_QUERY_ARGUMENT_NAME = 'queryArgument';
+
+const kotlinOperator: Dictionary<string> = {
+  [Op.AND]: '&&',
+  [Op.OR]: '||',
+  [Op.LT]: '<',
+  [Op.GT]: '>',
+  [Op.LTE]: '<=',
+  [Op.GTE]: '>=',
+  [Op.ADD]: '+',
+  [Op.SUB]: '-',
+  [Op.MUL]: '*',
+  [Op.DIV]: '/',
+  [Op.NOT]: '!',
+  [Op.NEG]: '-',
+  [Op.EQ]: '==',
+  [Op.NEQ]: '!=',
+};
+
+class KotlinRefinementGenerator extends RefinementExpressionVisitor<string> {
+
+  visitBinaryExpression(expr: BinaryExpression): string {
+    return `(${this.visit(expr.leftExpr)} ${kotlinOperator[expr.operator.op]} ${this.visit(expr.rightExpr)})`;
+  }
+  visitUnaryExpression(expr: UnaryExpression): string {
+    return `(${kotlinOperator[expr.operator.op]}${this.visit(expr.expr)})`;
+  }
+  visitFieldNamePrimitive(expr: FieldNamePrimitive): string {
+    return (expr.value.toString());
+  }
+  visitQueryArgumentPrimitive(_: QueryArgumentPrimitive): string {
+    return KOTLIN_QUERY_ARGUMENT_NAME;
+  }
+  visitBuiltIn(expr: BuiltIn): string {
+    // TODO: Double check that millis are the correct default units.
+    if (expr.value === 'now()') {
+      return `System.currentTimeMillis()`;
+    }
+
+    // TODO: Implement KT getter for 'creationTimeStamp'
+    throw new Error(`Unhandled BuiltInNode '${expr.value}' in toKTExpression`);
+  }
+  visitNumberPrimitive(expr: NumberPrimitive): string {
+    // This assumes that the associated Kotlin type will be `double`.
+    if (expr.value === Infinity) {
+      return 'Double.POSITIVE_INFINITY';
+    }
+    if (expr.value === -Infinity) {
+        return 'Double.NEGATIVE_INFINITY';
+    }
+    return expr.value.toString();
+  }
+  visitBooleanPrimitive(expr: BooleanPrimitive): string {
+    return `${expr.value}`;
+  }
+  visitTextPrimitive(expr: TextPrimitive): string {
+    // TODO(cypher1): Consider escaping this for Kotlin code generation.
```

Comment:
This also seems like an important TODO(b/) that needs to be filed.

Reply:
Filed b/159211498 against Josh for the Kotlin escaping, placed the b/ in code and I've put it on the tracking dashboard.